### PR TITLE
feat: ❇️ Provide tooltip action buttons

### PR DESCRIPTION
**Pull request by harsh-u-simform**
# Description
This PR allows us to provide the list of tooltip action button with configurations.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have followed the [Contributor Guide] when preparing my PR.
- [  ] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

[Contributor Guide]: https://github.com/SimformSolutionsPvtLtd/flutter_showcaseview/blob/master/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org

## Video of the new feature
[379697654-ed82cee2-f283-49e8-bfe5-a79f66e151eb.webm](https://github.com/user-attachments/assets/66168d9e-e7fe-41a6-a6fd-b496700af501)

### CODE REVIEW

File: example/lib/main.dart [hunk 1]
```diff
@@ -221,7 +236,34 @@ class _MailPageState extends State<MailPage> {
                           "Tap to see profile which contains user's name, profile picture, mobile number and country",
                       tooltipBackgroundColor: Theme.of(context).primaryColor,
                       textColor: Colors.white,
+                      onTargetClick: () {
```
Comment (vatsaltanna-simformsolutions) [inline example/lib/main.dart:239]: remove unrelated changes from this PR.

File: example/lib/main.dart [hunk 2]
```diff
@@ -311,27 +382,57 @@ class _MailPageState extends State<MailPage> {
...
+                type: TooltipDefaultActionType.previous,
+                name: 'Back',
+                onTap: () {
+                  // Write your code on button tap
+                  ShowCaseWidget.of(context).previous();
+                },
+                backgroundColor: Colors.pink.shade50,
+                textStyle: const TextStyle(
+                  color: Colors.pink,
+                )),
```
Comment (vatsaltanna-simformsolutions) [inline example/lib/main.dart:405]: format the code.

File: lib/src/showcase.dart [hunk 3]
```diff
@@ -26,11 +26,10 @@ import 'dart:ui';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 
-import 'enum.dart';
+import '../showcaseview.dart';
```
Comment (vatsaltanna-simformsolutions) [inline lib/src/showcase.dart:29]: any reason for this import?

File: lib/src/showcase.dart [hunk 4]
```diff
@@ -299,12 +309,14 @@ class Showcase extends StatefulWidget {
     this.tooltipPosition,
     this.titlePadding,
     this.descriptionPadding,
+    this.tooltipActions,
```
Comment (vatsaltanna-simformsolutions) [inline lib/src/showcase.dart:312]: put action and config together in constructor

File: lib/src/showcase.dart [hunk 5]
```diff
@@ -641,11 +655,60 @@ class _ShowcaseState extends State<Showcase> {
...
   }
+
+  List<Widget> _getTooltipActions() {
+    final showCaseState = ShowCaseWidget.of(context);
+    final actionData = (widget.tooltipActions?.isEmpty ?? true)
+        ? showCaseState.globalTooltipActions ?? []
+        : widget.tooltipActions ?? [];
+
+    final actionWidgets = <Widget>[];
+    for (var action = 0; action < actionData.length; action++) {
```
Comment (vatsaltanna-simformsolutions) [inline lib/src/showcase.dart:673]: should use for in loop

File: lib/src/showcase.dart [hunk 6]
```diff
@@ -641,11 +655,60 @@ class _ShowcaseState extends State<Showcase> {
...
+            tooltipActionConfig: _getTooltipActionConfig(),
+            tooltipActions: _getTooltipActions(),
           ),
         ],
       ],
     );
   }
+
+  List<Widget> _getTooltipActions() {
+    final showCaseState = ShowCaseWidget.of(context);
```
Comment (vatsaltanna-simformsolutions) [inline lib/src/showcase.dart:667]: use context check before this

File: lib/src/showcase.dart (hunk 3 above)
Comment (Sahil-Simform) [reply lib/src/showcase.dart:29]: No particular reason for this import; it just reduces the number of other imports needed.

Removing this and adding individual file imports.

File: lib/src/showcase.dart (hunk 6 above)
Comment (Sahil-Simform) [reply lib/src/showcase.dart:667]: It is used in the build method to provide action for the widget, eliminating the need for a context check since the context will always be mounted in the build method.

File: lib/src/showcase.dart [hunk 7]
```diff
@@ -641,11 +655,60 @@ class _ShowcaseState extends State<Showcase> {
...
+      actionWidgets.add(
+        Padding(
+          padding: EdgeInsetsDirectional.only(
+            end: action < actionData.length - 1
+                ? _getTooltipActionConfig().actionGap
+                : 0,
+          ),
+          child: TooltipActionButtonWidget(
+            config: actionData[action],
+            showCaseState: ShowCaseWidget.of(context),
```
[the remark this answers is not in the document]
Comment (Sahil-Simform) [reply lib/src/showcase.dart:694]: ToolTip is not a child of the ShowCaseWidget, so we won't get the `showCaseState` directly in the tooltip action using the context. Therefore, we passed that as an argument.

image

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,7 @@
+## [4.0.0]
+
+- Feature ✨: Added Action widget for tooltip
+
 ## [3.0.1]
 
 - Feature [#475](https://github.com/SimformSolutionsPvtLtd/flutter_showcaseview/issues/475) - Add
```
[the remark this answers is not in the document]
Comment (Sahil-Simform) [reply CHANGELOG.md:3]: Updated

File: lib/src/widget/action_widget.dart [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox(
+      width: width,
+      child: Padding(
+        padding: EdgeInsets.only(
+          top: isArrowUp ? tooltipActionConfig.gapBetweenContentAndAction : 0.0,
+          bottom:
+              !isArrowUp ? tooltipActionConfig.gapBetweenContentAndAction : 0.0,
+        ).add(outSidePadding),
```
Comment (ujas-m-simformsolutions) [inline lib/src/widget/action_widget.dart:34]: Do add operation in build.

File: lib/src/tooltip_widget.dart [hunk 10]
```diff
@@ -154,26 +200,76 @@ class _ToolTipWidgetState extends State<ToolTipWidget>
...
+    // adjust it to stay in available size
+    if (maxTextWidth > availableSpaceForToolTip) {
+      tooltipWidth = availableSpaceForToolTip;
     } else {
-      tooltipWidth = maxTextWidth + tooltipTextPadding;
+      // This is padding we will have around the tooltip text
+      final textPadding =
+          (widget.tooltipPadding ?? EdgeInsets.zero).horizontal +
+              max((widget.titlePadding ?? EdgeInsets.zero).horizontal,
+                  (widget.descriptionPadding ?? EdgeInsets.zero).horizontal);
```
Comment (ujas-m-simformsolutions) [inline lib/src/tooltip_widget.dart:236]: at this point just create a variable named zeroPadding.

File: lib/src/showcase.dart [hunk 11]
```diff
@@ -641,11 +775,54 @@ class _ShowcaseState extends State<Showcase> {
...
+      /// then it will hide that action for current widget
+      if (_showShowCase &&
+          action.hideActionForKeys.contains(widget.key) &&
+          (widget.tooltipActions?.isEmpty ?? true)) {
+        continue;
+      }
+      actionWidgets.add(
+        Padding(
+          padding: EdgeInsetsDirectional.only(
+            end: action != actionData.last
```
Comment (ujas-m-simformsolutions) [inline lib/src/showcase.dart:804]: I think you have to override `==` operator for `TooltipActionConfig`

**Review by ujas-m-simformsolutions — CHANGES_REQUESTED** [review]

File: lib/src/tooltip_widget.dart [hunk 12]
```diff
@@ -316,20 +413,26 @@ class _ToolTipWidgetState extends State<ToolTipWidget>
   @override
   void didChangeDependencies() {
     super.didChangeDependencies();
-    _getTooltipWidth();
+    // If tooltip is dismissing then no need to recalculate the size and widgets
+    if (!widget.isTooltipDismissed) {
```
[the remark this answers is not in the document]
Comment (Sahil-Simform) [reply lib/src/tooltip_widget.dart:417]: We do not have oldWidget here so won't be able to compare old and new widget hash.

File: lib/src/tooltip_widget.dart [hunk 13]
```diff
@@ -533,21 +723,74 @@ class _ToolTipWidgetState extends State<ToolTipWidget>
...
   }
 
-  Size _textSize(String text, TextStyle style) {
+  Size? _textSize(String? text, TextStyle style, EdgeInsets? padding) {
+    if (text == null) {
+      return null;
+    }
+
+    final availableSpaceForText =
+        (widget.position?.screenWidth ?? MediaQuery.of(context).size.width) -
```
[the remark this answers is not in the document]
Comment (Sahil-Simform) [reply lib/src/tooltip_widget.dart:776]: We support Flutter version up to 3.3.0, and sizeOf was released afterward, so it cannot be used.

File: lib/src/tooltip_widget.dart [hunk 14]
```diff
@@ -518,9 +676,41 @@ class _ToolTipWidgetState extends State<ToolTipWidget>
...
+                        children: [
+                          Stack(
+                            children: [
+                              // This widget is used for calculation of the action
+                              // widget size and it will be removed once the size
+                              // is calculated
+                              // We have kept it in colum because if we put is
+                              // outside in the stack then it will take whole
+                              // screen size and width calculation will fail
+                              if (isSizeRecalculating) _getOffstageActionWidget,
```
[the remark this answers is not in the document]
Comment (Sahil-Simform) [reply lib/src/tooltip_widget.dart:689]: Yes this Widget will stay offStage and will be used for only one frame to calculate size of action and then it will be discarded so this is correct place for this.

File: lib/src/tooltip_widget.dart [hunk 15]
```diff
@@ -342,19 +445,28 @@ class _ToolTipWidgetState extends State<ToolTipWidget>
         contentOrientation == TooltipPosition.bottom ? 1.0 : -1.0;
     isArrowUp = contentOffsetMultiplier == 1.0;
 
-    final contentY = isArrowUp
+    final screenSize = MediaQuery.of(context).size;
```
[the remark this answers is not in the document]
Comment (Sahil-Simform) [reply lib/src/tooltip_widget.dart:448]: We support Flutter version up to 3.3.0, and sizeOf was released afterward, so it cannot be used.

File: lib/src/showcase.dart (hunk 11 above)
Comment (Sahil-Simform) [reply lib/src/showcase.dart:804]: As discussed this logic is working properly so there is no need to override ```==``` method

**Review by ujas-m-simformsolutions — APPROVED** [review]

File: lib/src/tooltip_action_button_widget.dart [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  /// showcase navigation
+  final ShowCaseWidgetState showCaseState;
+
+  @override
+  Widget build(BuildContext context) {
+    final theme = Theme.of(context);
+
+    return config.button ??
+        GestureDetector(
+          onTap: handleOnTap,
```
[the remark this answers is not in the document]
Comment (Sahil-Simform) [reply lib/src/tooltip_action_button_widget.dart:25]: Yes it will work on transparent part for that user has to provide the padding as needed.
Updated Example application for this.

**Review by JanakMistry2000 — APPROVED** [review]
